metronome/checkpointing/service/test/src/io/iohk/metronome/checkpointing/service/CheckpointingServiceProps.scala

```diff
@@ -0,0 +1,407 @@
+          )
+
+        } yield TestResources(
+          service,
+          ledgerStorage,
+          blockStorage,
+          store,
+          ledgerTree
+        )
+      }
+    }
+
+    // not used in the impl so a senseless value
+    val commitPath = NonEmptyList.one(initialBlock.header.parentHash)
+
+    val allTransactions = batch.flatMap(_.body.transactions)
+    val finalLedger     = initialLedger.update(allTransactions)
+  }
+
+  object TestFixture {
+    import ArbitraryInstances._
+
+    def gen(minChain: Int = 1): Gen[TestFixture] = {
+      for {
+        block <- arbitrary[Block]
+        ledger = Ledger.empty.update(block.body.transactions)
+        batch    <- genBlockChain(block, ledger, min = minChain)
+        commitQC <- genCommitQC(batch.last)
+      } yield TestFixture(block, ledger, batch, commitQC)
+    }
+
+    def genBlockChain(
+        parent: Block,
+        initialLedger: Ledger,
+        min: Int = 1,
+        max: Int = 6
+    ): Gen[List[Block]] = {
+      for {
+        n      <- Gen.choose(min, max)
+        blocks <- Gen.listOfN(n, arbitrary[Block])
+      } yield {
+        def link(
+            parent: Block,
+            prevLedger: Ledger,
+            chain: List[Block]
+        ): List[Block] = chain match {
+          case b :: bs =>
+            val nextLedger = prevLedger.update(b.body.transactions)
+            val header = b.header.copy(
+              parentHash = parent.hash,
+              height = parent.header.height + 1,
+              postStateHash = nextLedger.hash
+            )
+            val linked = Block.makeUnsafe(header, b.body)
+            linked :: link(linked, nextLedger, bs)
+          case Nil =>
+            Nil
+        }
+
+        link(parent, initialLedger, blocks)
+      }
+    }
+
+    def genCommitQC(
+        block: Block
+    ): Gen[QuorumCertificate[CheckpointingAgreement]] =
+      arbitrary[QuorumCertificate[CheckpointingAgreement]].map {
+        _.copy[CheckpointingAgreement](phase = Commit, blockHash = block.hash)
+      }
+  }
+
+  def run(fixture: TestFixture)(test: TestResources => Task[Prop]): Prop = {
+    import Scheduler.Implicits.global
+
+    fixture.resources.use(test).runSyncUnsafe(timeout = 5.seconds)
+  }
+
+  property("normal execution") = forAll(TestFixture.gen()) { fixture =>
+    run(fixture) { res =>
+      import fixture._
+      import res._
+
+      val execution = batch
+        .map(checkpointingService.executeBlock(_, commitQC, commitPath))
+        .sequence
+
+      val ledgerStorageCheck = store.runReadOnly {
+        ledgerStorage.get(finalLedger.hash)
+      }
+
+      for {
+        results         <- execution
+        persistedLedger <- ledgerStorageCheck
+        ledgerTree      <- ledgerTreeRef.get
+      } yield {
+        val ledgerTreeUpdated = ledgerTree == Map(
+          batch.last.hash -> (finalLedger, batch.last.header)
+        )
+
+        all(
+          "executed correctly" |: results.reduce(_ && _),
```

Comment:
By "executed correctly" you mean that "executed according to expectations", which is that all executions returned `true`, not that a `false` would signal incorrect execution, right? 

Just wondering if something like "persisted all executions" would make it clearer.

Reply:
There is a separate check for persisting the state. Maybe simply "executed as expected"?